### PR TITLE
executor: make IndexReaders in the inner side of IndexJoins support direct reading

### LINKED ISSUE #24255: Make IndexReaders in the inner side of IndexJoins support direct reading
## Development Task
`IndexReader` already has the ability to read partition tables directly, but if it is in the inner side of an `IndexJoin`, `PartitionTableExec` is still used now, and this is unnecessary.
To make our code more uniform, we should remove `PartitionTableExec` in this case and just use the ability of direct reading of `IndexReader`.

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #24255 

Problem Summary: executor: make IndexReaders in the inner side of IndexJoins support direct reading

### What is changed and how it works?

Please see more details at https://github.com/pingcap/tidb/issues/24255.
In this PR, we:
1. use `IndexReader` directly instead of using `PartitionExecForIndexReader` to read partition table;
2. remove `PartitionExecForIndexReader`;
3. add a test case;

### Check List 

Tests 

- Unit test

### Release note 

- executor: make IndexReaders in the inner side of IndexJoins support direct reading

### CODE REVIEW

**Comment by qw4990** [comment]
This PR is small, please help us review this PR at your convenience. @lysu @tiancaiamao Very appreciate!

**Comment by qw4990** [comment]
Could you please help us review this small PR? @wshwsh12  Very appreciate it!

File: executor/index_lookup_join_test.go [hunk 1]
```diff
@@ -331,3 +334,35 @@ func (s *testSuite5) TestIssue23722(c *C) {
...
+func (s *testSuite5) TestPartitionTableIndexJoinAndIndexReader(c *C) {
+	if israce.RaceEnabled {
+		c.Skip("exhaustive types test, skip race test")
+	}
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("set @@tidb_partition_prune_mode='dynamic'")
+	tk.MustExec(`create table t (a int, b int, key(a)) partition by hash(a) partitions 4`)
+	tk.MustExec("create table tnormal (a int, b int, key(a), key(b))")
+	nRows := 64
```
Comment (wshwsh12) [inline executor/index_lookup_join_test.go:347]: I try to increase nRows to larger data, for example nRows=640.. and test again. It wil panic.

Comment (wshwsh12) [reply executor/index_lookup_join_test.go:347]:
```
index_lookup_join_test.go:366:
    tk.MustQuery("select /*+ TIDB_INLJ(t1, t2) */ t1.a from t t1, t t2 where t1.a=t2.b and " + cond).Sort().Check(result)
/home/wshwsh12/project/tidb/util/testkit/testkit.go:346:
    tk.c.Check(errors.ErrorStack(err), check.Equals, "", comment)
... obtained string = "" +
...     "runtime error: index out of range [-1]\n" +
...     "github.com/pingcap/tidb/executor.(*innerWorker).run.func1\n" +
...     "\t/home/wshwsh12/project/tidb/executor/index_lookup_join.go:454\n" +
...     "runtime.gopanic\n" +
...     "\t/usr/lib/go/src/runtime/panic.go:965\n" +
...     "runtime.goPanicIndex\n" +
...     "\t/usr/lib/go/src/runtime/panic.go:88\n" +
...     "github.com/pingcap/parser.(*Scanner).stmtText\n" +
...     "\t/home/wshwsh12/go/pkg/mod/github.com/pingcap/parser@v0.0.0-20210330190622-f959a136fc19/lexer.go:96\n" +
...     "github.com/pingcap/parser.yyParse\n" +
...     "\t/home/wshwsh12/go/pkg/mod/github.com/pingcap/parser@v0.0.0-20210330190622-f959a136fc19/parser.go:17874\n" +
...     "github.com/pingcap/parser.(*Parser).Parse\n" +
...     "\t/home/wshwsh12/go/pkg/mod/github.com/pingcap/parser@v0.0.0-20210330190622-f959a136fc19/yy_parser.go:152\n" +
...     "github.com/pingcap/tidb/session.(*session).ParseSQL\n" +
...     "\t/home/wshwsh12/project/tidb/session/session.go:1091\n" +
...     "github.com/pingcap/tidb/expression.ParseSimpleExprsWithNames\n" +
...     "\t/home/wshwsh12/project/tidb/expression/simple_rewriter.go:88\n" +
...     "github.com/pingcap/tidb/planner/core.generateHashPartitionExpr\n" +
...     "\t/home/wshwsh12/project/tidb/planner/core/rule_partition_processor.go:110\n" +
...     "github.com/pingcap/tidb/planner/core.(*partitionProcessor).findUsedPartitions\n" +
...     "\t/home/wshwsh12/project/tidb/planner/core/rule_partition_processor.go:121\n" +
...     "github.com/pingcap/tidb/planner/core.(*partitionProcessor).pruneHashPartition\n" +
...     "\t/home/wshwsh12/project/tidb/planner/core/rule_partition_processor.go:207\n" +
...     "github.com/pingcap/tidb/planner/core.PartitionPruning\n" +
...     "\t/home/wshwsh12/project/tidb/planner/core/partition_prune.go:38\n" +
...     "github.com/pingcap/tidb/executor.partitionPruning\n" +
...     "\t/home/wshwsh12/project/tidb/executor/builder.go:3986\n" +
...     "github.com/pingcap/tidb/executor.prunePartitionForInnerExecutor\n" +
...     "\t/home/wshwsh12/project/tidb/executor/builder.go:2780\n" +
...     "github.com/pingcap/tidb/executor.(*dataReaderBuilder).buildIndexReaderForIndexJoin\n" +
...     "\t/home/wshwsh12/project/tidb/executor/builder.go:3478\n" +
...     "github.com/pingcap/tidb/executor.(*dataReaderBuilder).buildExecutorForIndexJoinInternal\n" +
...     "\t/home/wshwsh12/project/tidb/executor/builder.go:3211\n" +
...     "github.com/pingcap/tidb/executor.(*dataReaderBuilder).buildExecutorForIndexJoin\n" +
...     "\t/home/wshwsh12/project/tidb/executor/builder.go:3202\n" +
...     "github.com/pingcap/tidb/executor.(*innerWorker).fetchInnerResults\n" +
...     "\t/home/wshwsh12/project/tidb/executor/index_lookup_join.go:647\n" +
...     "github.com/pingcap/tidb/executor.(*innerWorker).handleTask\n" +
...     "\t/home/wshwsh12/project/tidb/executor/index_lookup_join.go:491\n" +
...     "github.com/pingcap/tidb/executor.(*innerWorker).run\n" +
...     "\t/home/wshwsh12/project/tidb/executor/index_lookup_join.go:469\n" +
...     "runtime.goexit\n" +
...     "\t/usr/lib/go/src/runtime/asm_amd64.s:1371"
... expected string = ""
... sql:select /*+ TIDB_INLJ(t1, t2) */ t1.a from t t1, t t2 where t1.a=t2.b and (t2.b between 304 and 559), args:[]
```

Comment (qw4990) [reply executor/index_lookup_join_test.go:347]: Yes, this is a known problem, and actually, it's another bug irrelevant to this PR.
When it is in static mode (`set @@tidb_partition_prune_mode='dynamic'`), this problem still exists.
I have already created an issue for it https://github.com/pingcap/tidb/issues/24259, and we decide to solve it later in our testing and bugfix phase.
PTAL @wshwsh12

**Comment by qw4990** [comment]
PTAL @tiancaiamao

File: executor/index_lookup_join_test.go (hunk 1 above)
Comment (tiancaiamao) [reply executor/index_lookup_join_test.go:347]: Seems a concurrency bug.
Partition pruning use a parser shared in the session.

**Pull request merged**